controllers/agentmachine_controller.go

```diff
@@ -141,51 +145,84 @@ func (r *AgentMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request
-		// Register a finalizer if it is absent.
-		if !funk.ContainsString(agentMachine.GetFinalizers(), AgentMachineFinalizerName) {
-			controllerutil.AddFinalizer(agentMachine, AgentMachineFinalizerName)
-			if err := r.Update(ctx, agentMachine); err != nil {
-				log.WithError(err).Errorf("failed to add finalizer %s to resource %s %s", AgentMachineFinalizerName, agentMachine.Name, agentMachine.Namespace)
-				return &ctrl.Result{}, err
-			}
+func (r *AgentMachineReconciler) removeMachineDeletionHookAnnotation(ctx context.Context, machine *clusterv1.Machine) (err error) {
+	annotations := machine.GetAnnotations()
+	if _, haveMachineHookAnnotation := annotations[machineDeleteHookName]; haveMachineHookAnnotation {
+		delete(annotations, machineDeleteHookName)
+		machine.SetAnnotations(annotations)
+		err = r.Update(ctx, machine)
+	}
+	return err
+}
+
+func (r *AgentMachineReconciler) handleDeletionHook(ctx context.Context, log logrus.FieldLogger, agentMachine *capiproviderv1alpha1.AgentMachine, machine *clusterv1.Machine) (*ctrl.Result, error) {
+	// TODO: this can be removed when we're sure no agent machines have this finalizer anymore
+	if funk.ContainsString(agentMachine.GetFinalizers(), AgentMachineFinalizerName) {
+		controllerutil.RemoveFinalizer(agentMachine, AgentMachineFinalizerName)
+		if err := r.Update(ctx, agentMachine); err != nil {
+			log.WithError(err).Errorf("failed to remove finalizer %s from resource %s %s", AgentMachineFinalizerName, agentMachine.Name, agentMachine.Namespace)
+			return &ctrl.Result{}, err
 		}
-	} else { // AgentMachine is being deleted
-		r.Log.Info("Found deletion timestamp on AgentMachine")
-		if funk.ContainsString(agentMachine.GetFinalizers(), AgentMachineFinalizerName) {
-			// deletion finalizer found, unbind the Agent from the ClusterDeployment
-			if agentMachine.Status.AgentRef != nil {
-				r.Log.Info("Removing ClusterDeployment ref to unbind Agent")
-				agent, err := r.getAgent(ctx, log, agentMachine)
-				if err != nil {
-					if apierrors.IsNotFound(err) {
-						log.WithError(err).Infof("Failed to get agent %s. assuming the agent no longer exists", agentMachine.Status.AgentRef.Name)
-					} else {
-						log.WithError(err).Errorf("Failed to get agent %s", agentMachine.Status.AgentRef.Name)
-						return &ctrl.Result{}, err
-					}
-				} else {
-					// Remove the AgentMachineRefLabel and set the clusterDeployment to nil
-					delete(agent.ObjectMeta.Labels, AgentMachineRefLabelKey)
-					agent.Spec.ClusterDeploymentName = nil
-					if err := r.Update(ctx, agent); err != nil {
-						log.WithError(err).Error("failed to remove the Agent's ClusterDeployment ref")
-						return &ctrl.Result{}, err
-					}
-				}
-			}
+	}
+
+	// set delete hook if not present
+	annotations := machine.GetAnnotations()
+	if _, haveMachineHookAnnotation := annotations[machineDeleteHookName]; !haveMachineHookAnnotation {
+		if annotations == nil {
+			annotations = make(map[string]string)
+		}
```

Comment:
The order seems off

Reply:
Nope, this is okay. We only need to make the map when we want to assign to it which is in this condition.

It's okay to check for the presence of an entry in a nil map. It will just return `false` for anything.